contracts/anchor_handler/lib.rs

```diff
@@ -0,0 +1,272 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use ink_lang as ink;
+
+#[ink::contract]
+mod anchor_handler {
+    use ink_prelude::string::String;
+    use ink_prelude::vec::Vec;
+    use ink_storage::traits::{PackedLayout, SpreadLayout, StorageLayout};
+    use ink_storage::{traits::SpreadAllocate, Mapping};
+    use protocol_ink_lib::keccak::Keccak256;
+    use protocol_ink_lib::utils::{element_encoder, element_encoder_for_four_bytes};
+    use vanchor::vanchor::TokenWrapperData;
+    use vanchor::VAnchorRef;
+    /// The token wrapper handler result type.
+    pub type Result<T> = core::result::Result<T, Error>;
+
+    #[ink(storage)]
+    #[derive(SpreadAllocate)]
+    pub struct AnchorHandler {
+        /// Contract address of previously deployed Bridge.
+        bridge_address: AccountId,
+        /// resourceID => token contract address
+        resource_id_to_contract_address: Mapping<[u8; 32], AccountId>,
+        /// Execution contract address => resourceID
+        contract_address_to_resource_id: Mapping<AccountId, [u8; 32]>,
+        /// Execution contract address => is whitelisted
+        contract_whitelist: Mapping<AccountId, bool>,
+        /// (src_chain_id, height) -> UpdateRecord
+        update_records: Mapping<(u64, u64), UpdateRecord>,
+        vanchor: VAnchorRef,
+    }
+
+    #[derive(Default, Debug, scale::Encode, scale::Decode, Clone, SpreadLayout, PackedLayout)]
+    #[cfg_attr(feature = "std", derive(StorageLayout, scale_info::TypeInfo))]
+    pub struct UpdateRecord {
+        pub token_address: AccountId,
+        pub src_chain_id: u64,
+        pub resource_id: [u8; 32],
+        pub merkle_root: [u8; 32],
+        pub leaf_id: u64,
```

Comment:
Shouldn’t this be a u32?

Reply:
protcol-cosmwasm represented a u64 though https://github.com/webb-tools/protocol-cosmwasm/blob/main/contracts/anchor-handler/src/state.rs#L13